agent/xds/clusters.go

```diff
@@ -133,32 +133,56 @@ func (s *Server) clustersFromSnapshotMeshGateway(cfgSnap *proxycfg.ConfigSnapsho
 		if dc == cfgSnap.Datacenter {
 			continue // skip local
 		}
-		clusterName := connect.DatacenterSNI(dc, cfgSnap.Roots.TrustDomain)
+		useEDS := true
+		if len(cfgSnap.MeshGateway.HostnameDatacenters[dc]) > 0 {
+			useEDS = false
+		}
 
-		cluster, err := s.makeGatewayCluster(cfgSnap, clusterName)
+		cluster := &envoy.Cluster{
+			Name: connect.DatacenterSNI(dc, cfgSnap.Roots.TrustDomain),
+		}
+		err := s.populateGatewayCluster(cfgSnap, cluster, 0, useEDS)
 		if err != nil {
 			return nil, err
 		}
+		if !useEDS {
+			injectHostnameEndpoints(cluster, cfgSnap.MeshGateway.HostnameDatacenters[dc], dc != cfgSnap.Datacenter, false)
+		}
+
 		clusters = append(clusters, cluster)
 	}
 
 	if cfgSnap.ServiceMeta[structs.MetaWANFederationKey] == "1" && cfgSnap.ServerSNIFn != nil {
 		// Add all of the remote wildcard datacenter mappings for servers.
 		for _, dc := range datacenters {
-			clusterName := cfgSnap.ServerSNIFn(dc, "")
+			if dc == cfgSnap.Datacenter {
```

Comment:
This section is about setting up the wildcard clusters for getting OUT to another datacenter `(dc != cfgSnap.Datacenter)` and *also* one for getting IN from another datacenter `(dc == cfgSnap.Datacenter)`.

You should not omit the local dc here, but you still should handle it differently now. When `dc == cfgSnap.Datacenter` it should behave like before, since it's definitely not going to be passing hostnames here, since we're using our real local catalog entries which are going to have full IP addresses for our own servers.

Reply:
Ah, yep that fixed it.